### PR TITLE
Fix missing initial and deletion notification flow events

### LINKED ISSUE #1151: Missing InitialObject event on BaseRealmObject.asFlow
`BaseRealmObject.asFlow()` does not always emit `InitialObject` as the first event. As part of https://github.com/realm/realm-kotlin/pull/1132/files#diff-4a4a616396d054e630f87ed03cebff2cf991671f96ef6193d3f92aa89bd18e8bR505 the amount of updates during registration seems to trigger that the first [`onChange`](https://github.com/realm/realm-kotlin/blob/master/packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/SuspendableNotifier.kt#L104) callback on a later version that the one where we register for notifications
```
register: io.realm.kotlin.internal.RealmObjectReference@81bb19a VersionId(version=3) VersionId(version=9)
onChange: io.realm.kotlin.internal.RealmObjectReference@81bb19a VersionId(version=10)
flow: io.realm.kotlin.notifications.internal.UpdatedObjectImpl@cc23154 VersionId(version=10) 
```
This means that [this assumption](https://github.com/realm/realm-kotlin/blob/main/packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmObjectReference.kt#L132) isn't right, since the first change event actually holds changes and will not report an initial `InitialObject` event. 

Probably also the case for the similar assumption for emitting initial events on lists, results and sets, since it uses the same infrastructure/logic.

### LINKED ISSUE #1233: Cannot listen for changes on deleted object
If registering for notification on a deleted entity the `collect { ... }` will throw `java.lang.IllegalStateException: Cannot listen for changes on a deleted Realm reference`. It should probably just emit a `DeletedObject`-change event and complete the flow. 

```
val obj: Sample = realm.write { copyToRealm(Sample().apply { stringField = "Foo" }) }
realm.write { delete(findLatest(obj)!!) }

obj.asFlow().collect { }
```

### PR DESCRIPTION
**Pull request by rorbech**
This PR ensures that we always get an _initial_-event from notification flows with the state of the realm when the actual listener is register with the notifier thread. The PR also fixes behavior when registering for notification on outdated entities. This is now updated to match the behavior for deletion of entities on running flows, thus emitting a _deletion_-event and completing the flow gracefully, as opposed to throwing a `RuntimeException` from `error("Cannot listen for changes on a deleted Realm reference")`.

*Reviewer notes*

You should start out with reviewing the interfaces in https://github.com/realm/realm-kotlin/pull/1273/files#diff-5214e2ae8dd748b28d576217d538635b608b6c4fa617638bcbfa11474c9a5580R24 to get the best understanding of the changes

*TODOs*
- [x] Clean up
- [x] Internal doc update
- [x] Fix flaky backlink notification test

Closes #1151
Closes #1233

### CODE REVIEW

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt [hunk 1]
```diff
@@ -15,38 +15,118 @@
...
 
-// TODO Public due to being a transitive dependency to Observable
-public interface Freezable<T> {
-    public fun freeze(frozenRealm: RealmReference): T?
+/**
+ * A _notifiable_ yields the live reference and change event builder that is used by the
+ * [SuspendableNotifier] to register for notifications with core and convert the core change sets
+ * into [C]-change events.
+ */
+public interface Notifiable<T : CoreNotifiable<T, C>, C> {
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt:47]: this should be internal

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt [hunk 2]
```diff
@@ -15,38 +15,118 @@
...
 
-// TODO Public due to being a transitive dependency to Observable
-public interface Thawable<T> {
-    public fun thaw(liveRealm: RealmReference): T?
+/**
+ * A _change flow_ is responsible for converting the [SuspendableNotifier] events into
+ * corresponding [C]-change events and emit them in the [producerScope] and close the scope when
+ * the monitored entity is deleted.
+ */
+public abstract class ChangeFlow<T, C>(private val producerScope: ProducerScope<C>) {
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt:67]: this should be internal

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt [hunk 3]
```diff
@@ -15,38 +15,118 @@
...
-public interface Observable<T, C> : Notifiable<C>, Freezable<Observable<T, C>>, Thawable<Observable<T, C>>
+/**
+ * A _core notifiable_ that supports the various operations on the entity [T] to support
+ * registration with the [SuspendableNotifier]. This includes thawing the initial frozen version
+ * in the the notifiers live context, register for notifications and freezing the updated version
+ * before signaling it to the flow.
+ *
+ * All [CoreNotifiable] are themselves also [Notifiable] and [Observable].
+ */
+public interface CoreNotifiable<T, C> : Notifiable<T, C>, Observable<T, C>, Versioned, Flowable<C>
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt:123]: this should be internal

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt [hunk 4]
```diff
@@ -15,38 +15,118 @@
...
-// TODO Public due to being a transitive dependency to Observable
-public interface Notifiable<T> {
-    public fun registerForNotification(callback: Callback<RealmChangesPointer>): RealmNotificationTokenPointer
+/**
+ * An _observable_ is an entity that from a user perspective supports some kind of notification
+ * mechanism. This does not necessarily mean that you can listen for changes of the object itself,
+ * but could be updates of derived entities, i.e. observing a query will actually emit results.
+ */
+// TODO Public due to being a transitive dependency to CoreNotifiable
+public interface Observable<T : CoreNotifiable<T, C>, C> {
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt:33]: this should be internal

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt [hunk 5]
```diff
@@ -15,38 +15,118 @@
...
+            producerScope.close()
+        }
+    }
+
+    internal abstract fun initial(frozenRef: T): C
+    internal abstract fun update(frozenRef: T, change: RealmChangesPointer): C?
+    internal abstract fun delete(): C
 }
 
 // TODO Why is this flowable here?
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt:110]: This should be moved to a different package?

**Review by clementetb — COMMENTED** [review]
Looks solid, added some minor comments.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt (hunk 5 above)
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt]: Which package should i be in? `io.realm.kotlin.internal` seems quite appropriate to me ... but I have moved it to a separate file. Don't know if that changes anything though 🤷

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/query/SingleQuery.kt [hunk 6]
```diff
@@ -57,37 +57,36 @@ internal class SingleQuery<E : BaseRealmObject> constructor(
...
+            .map { resultChange: ResultsChange<E> -> resultChange.list.firstOrNull() }
+            // Only react when head is changed
+            .distinctUntilChangedBy { head: BaseRealmObject? ->
+                head?.runIfManaged { RealmInterop.realm_object_get_key(this.objectPointer) }
+            }
+            // When head is changed issue flow that emits object changes
+            .flatMapLatest { newHead ->
+                val oldHeadDeleted =
+                    oldHead != null && (
+                        newHead == null || RealmInterop.realm_object_resolve_in(
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/query/SingleQuery.kt]: I have tried to update it, but I can't format it to something that makes sense for me due to the ktlint-rules and I can't find options to change indentation settings for sub groups of logical operations, so ... 🤷 ... If you have any recommended formatting I can update it with that.

[the remark this answers is not in the document]
Comment (edualonso) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/query/SingleQuery.kt]: You could extract the call to `realm_object_resolve_in` to a variable and that would most likely allow you to put everything in one line improving readability:
```
val objectRef = RealmInterop.realm_object_resolve_in(
    oldHead!!.realmObjectReference!!.objectPointer,
    newHead.realmObjectReference!!.owner.dbPointer
)
val oldHeadDeleted = oldHead != null && (newHead == null || objectRef == null)
```

File: packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/notifications/BacklinksNotificationsTests.kt [hunk 7]
```diff
@@ -333,8 +394,13 @@ class BacklinksNotificationsTests : NotificationTests {
...
                 )
             }
-            assertEquals(1, c.receive())
+
+            // Await that collect is actually collecting
+            withTimeout(10.seconds) {
+                assertEquals(1, c.receive())
+            }
             realm.close()
+            delay(1.seconds)
```
Comment (cmelchior) [inline packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/notifications/BacklinksNotificationsTests.kt:403]: 🙈 Lets hope this doesn't turn flaky

**Review by cmelchior — COMMENTED** [review]
This looks fine. I have one question about storing old references to objects though.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt [hunk 8]
```diff
@@ -15,38 +15,124 @@
...
+ * corresponding [C]-change events and emit them in the [producerScope] and close the scope when
+ * the monitored entity is deleted.
+
+ * @param T the type of entity that is observed.
+ * @param C the type of change events emitted for the T entity.
+ */
+// TODO Public due to being a transitive dependency from RealmObjectReference
+public abstract class ChangeFlow<T, C>(private val producerScope: ProducerScope<C>) {
+
+    private var previousElement: T? = null
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt]: Good point. I have updated it to be a boolean `initialEvent` flag.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmResultsImpl.kt [hunk 9]
```diff
@@ -121,25 +121,30 @@ internal class RealmResultsImpl<E : BaseRealmObject> constructor(
...
+    override fun update(
+        frozenRef: RealmResultsImpl<E>,
+        change: RealmChangesPointer
+    ): ResultsChange<E> {
+        val listChangeSetBuilderImpl = ListChangeSetBuilderImpl(change)
+        return UpdatedResultsImpl(frozenRef, listChangeSetBuilderImpl.build())
+    }
+
+    override fun delete(): ResultsChange<E> =
+        sdkError("Results should never have been deleted")
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmResultsImpl.kt:149]: `ResultsChange`s doesn't have a _deletion_ event, so we cannot emit specific events for that 🤷  We only complete the flow when the parent is deleted. This is implemented by backlinks being created as an `ObjectBoundRealmResults` from https://github.com/realm/realm-kotlin/blob/main/packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/BacklinksDelegateImpl.kt#L70 which monitors the parant object and closes the flow on deletion of that.

**Pull request merged**